website/integrations: add wazuh (#13776)
goauthentik/authentik: website/integrations/services/wazuh/index.mdx

support_level: community
---

## What is Wazuh

> Wazuh is a security platform that aims to safeguard data assets across multiple environments including on-site, virtualized, containerized, and cloud-based.
>
> -- https://wazuh.com

## Preparation

The following placeholders are used in this guide:

- `wazuh-dashboard.company` is the FQDN of the Wazuh Dashboard installation.
- `wazuh-indexer.company` is the FQDN of the Wazuh Indexer installation
- `authentik.company` is the FQDN of the authentik installation.

:::note
This documentation lists only the settings that you need to change from their default values. Be aware that any changes other than those explicitly mentioned in this guide could cause issues accessing your application.
:::

## authentik configuration

To support the integration of Wazuh with authentik, you need to create a group, a property mapping, and an application/provider pair in authentik.

### Create a user group in authentik

1. Log in to authentik as an admin, and open the authentik Admin interface.
2. Navigate to **Directory** > **Groups** and click **Create**.
3. Set a name for the group (e.g. `wazuh-administrators`) and click **Create**.

> 4. Click on the name of the newly created group and navigate to the **Users** tab.

Reviewer: ```suggestion
4. Click the name of the newly created group and navigate to the **Users** tab.
```

5. Click **Add existing user**, select the user that needs Wazuh admin access and click **Add**.

### Create a property mapping in authentik

1. Log in to authentik as an admin, and open the authentik Admin interface.
2. Navigate to **Customization** > **Property Mappings** and click **Create**. Create a **SAML Provider Property Mapping** with the following settings:

    - **Name**: Choose a descriptive name
    - **SAML Attribute Name**: <kbd>Roles</kbd>
    - **Friendly Name**: Leave blank
    - **Expression**: `wazuh-administrators` is used a placeholder for the group name.

> ```python

Reviewer: ```suggestion
     ```python
```

> if ak_is_group_member(request.user, name="wazuh-administrators"):

Reviewer: ```suggestion
     if ak_is_group_member(request.user, name="wazuh-administrators"):
```

>     yield "wazuh-admin"

Reviewer: ```suggestion
        yield "wazuh-admin"
```

> ```

Reviewer: ```suggestion
    ```
```

> Wazuh is a security platform that aims to safeguard data assets across multiple environments including on-site, virtualized, containerized, and cloud-based.
>
> -- https://github.com/wazuh/wazuh

## Preparation

The following placeholders are used in this guide:

- `wazuh-dashboard.company` is the FQDN of the Wazuh Dashboard installation.
- `wazuh-indexer.company` is the FQDN of the Wazuh Indexer installation
- `authentik.company` is the FQDN of the authentik installation.

:::note
This documentation lists only the settings that you need to change from their default values. Be aware that any changes other than those explicitly mentioned in this guide could cause issues accessing your application.
:::

## authentik configuration

To support the integration of Wazuh with authentik, you need to create a Wazuh user group, property mapping and application/provider pair in authentik.

### Create a user group in authentik

1. Log in to authentik as an admin, and open the authentik Admin interface.
2. Navigate to **Directory** -> **Groups** and click **Create** to create a group.
3. **Configure the group**:

    - **Name**: select a name for the group (e.g. `wazuh-administrators`).

> 4. Click **Create**

Author: I've merged steps 3 and 4. It's less repetitive now.

- **Configure Bindings** _(optional)_: you can create a [binding](/docs/add-secure-apps/flows-stages/bindings/) (policy, group, or user) to manage the listing and access to applications on a user's **My applications** page.

3. Click **Submit** to save the new application and provider.

### Download metadata file

1. Log in to authentik as an admin, and open the authentik Admin interface.
2. Navigate to **Applications** -> **Providers** and click on the name of the provider that you created in the previous section (e.g. `Provider for wazuh`).
3. Under **Related objects** -> **Metadata**, click on **Download**.

You will need this metadata file in the next section.

## Wazuh configuration

### Wazuh Indexer SAML configuration

1. For the next step you will require an exchange key, generate this key using this command:

    ```bash
    openssl rand -hex 32
    ```

2. Copy the metadata file (e.g. `wazuh_authentik_meta.xml`) that was downloaded in the previous section to the `/etc/wazuh-indexer/opensearch-security/` directory of your Wazuh Indexer server.

> 3. Set ownership of the file to `wazuh-indexer` using this command:

Author: It's difficult to know where to draw the line. I feel like we'll have issue reports if we don't mention it. Maybe we could just change it to an example?

:::

## authentik configuration

To support the integration of Wazuh with authentik, you need to create a group, a property mapping, and an application/provider pair in authentik.

### Create a user group in authentik

1. Log in to authentik as an admin, and open the authentik Admin interface.
2. Navigate to **Directory** > **Groups** and click **Create**.
3. Set a name for the group (e.g. `wazuh-administrators`) and click **Create**.
4. Click on the name of the newly created group and navigate to the **Users** tab.
5. Click on **Add existing user**, select the user that needs Wazuh admin access and click **Add**.

### Create a property mapping in authentik

1. Log in to authentik as an admin, and open the authentik Admin interface.
2. Navigate to **Customization** > **Property Mappings** and click **Create**. Create a **SAML Provider Property Mapping** with the following settings:

    - **Name**: Choose a descriptive name
    - **SAML Attribute Name**: <kbd>Roles</kbd>
    - **Friendly Name**: Leave blank
    - **Expression**:

<IntegrationsMultilineCodeblock>
    {`
if ak_is_group_member(request.user, name="<em>wazuh-administrators</em>"):
    yield "wazuh-admin"
`}

> </IntegrationsMultilineCodeblock>

Reviewer 2: ```suggestion
</IntegrationsMultilineCodeblock>;
```

Author: All these semicolons on the next line are down to prettier. It should be valid regardless.

> ;

Reviewer 2: ```suggestion
```

3. Click **Finish**.

### Create an application and provider in authentik

1. Log in to authentik as an admin, and open the authentik Admin interface.
2. Navigate to **Applications** > **Applications** and click **Create with Provider** to create an application and provider pair. (Alternatively you can first create a provider separately, then create the application and connect it with the provider.)

- **Application**: provide a descriptive name (e.g., `Wazuh`), an optional group for the type of application, the policy engine mode, and optional UI settings.
- **Choose a Provider type**: Select **SAML Provider** as the provider type.
- **Configure the Provider**: provide a name (or accept the auto-provided name), the authorization flow to use for this provider, and the following required configurations.
    - **ACS URL**: <kbd>https://<em>wazuh-dashboard.company</em>/\_opendistro/\_security/saml/acs</kbd>
    - **Issuer**: `wazuh-saml`
    - **Service Provider Binding**: `Post`

>     - **Property mappings**: add the property mapping created in previous section (e.g. `Wazuh Property Mapping`).

Reviewer 2: It's under advanced protocol settings. Also, I suggest:
```suggestion
    - - Under **Advanced protocol settings**, add the **Property Mapping** you created in the previous section.
```

>     - **NameID Property Mapping**: select a property mapping that will be used for Wazuh usernames (e.g. `authentik default SAML Mapping: Name` or `authentik default SAML Mapping: Email`)

Reviewer 2: Same with this, also under advanced protocol settings, we could perhaps merge both lines?

    - **Configure Bindings** _(optional)_: you can create a [binding](/docs/add-secure-apps/flows-stages/bindings/) (policy, group, or user) to manage the listing and access to applications on a user's **My applications** page.

3. Click **Submit** to save the new application and provider.

### Download metadata file

1. Log in to authentik as an admin, and open the authentik Admin interface.

> 2. Navigate to **Applications** -> **Providers** and click on the name of the provider that you created in the previous section (e.g. `Provider for wazuh`).

Reviewer 2: ```suggestion
2. Navigate to **Applications** > **Providers** and click on the name of the provider that you created in the previous section (e.g. `Provider for wazuh`).
```

> 3. Under **Related objects** -> **Metadata**, click on **Download**. This downloaded is your `SAML Metadata` file and it will be required in the next section.

Reviewer 2: ```suggestion
3. Under **Related objects** > **Metadata**, click on **Download**. The downloaded file is your **SAML Metadata** and will be required in the next section.
```

## Wazuh configuration

> 

Reviewer 2: Could we add an introduction here?

### Wazuh Indexer SAML configuration

:::note

> `wazuh_authentik_meta.xml` is used as a placeholder name for the `SAML Metadata` file. Ensure that the filename used in this section matches your `SAML Metadata` filename.

Reviewer 2: ```suggestion
The file `wazuh_authentik_meta.xml` serves as a placeholder for your SAML metadata file. Make sure that the filename used in this context matches the actual name of your SAML metadata file.
```

:::

1. For the next step, you will need an exchange key. To generate this key, use the following command:

    ```bash
    openssl rand -hex 32
    ```

> 2. Copy the metadata file (e.g., `wazuh_authentik_meta.xml`) downloaded in the previous step to the `/etc/wazuh-indexer/opensearch-security/` directory on your Wazuh Indexer server.

Reviewer 2: ```suggestion
2. Copy the downloaded metadata file to the `/etc/wazuh-indexer/opensearch-security/` directory on your Wazuh Indexer server.
```

> 3. Change the ownership of the `wazuh_authentik_meta.xml` file to the `wazuh-indexer` user using the following command:

Reviewer 2: ```suggestion
3. Change the ownership of the metadata file to the `wazuh-indexer` user using the following command:
```

Feels a little cheap to include the placeholder file name at every step, especially when the command is right below.

    ```bash
    chown wazuh-indexer:wazuh-indexer /etc/wazuh-indexer/opensearch-security/wazuh_authentik_meta.xml
    ```

> 4. Copy the metadata file (e.g., `wazuh_authentik_meta.xml`) downloaded in the previous step to the `/etc/wazuh-indexer/opensearch-security/` directory on your Wazuh Indexer server.

Reviewer 2: Same here

> import IntegrationsMultilineCodeblock from "@site/src/components/MultilineCodeblock";

Reviewer 2: Can we move this import to the top of the file?

<IntegrationsMultilineCodeblock>
    {`
---
authc:
    basic_internal_auth_domain:
        description: "Authenticate SAML against internal users database"
        http_enabled: true
        transport_enabled: true
        order: <em>0</em>
        http_authenticator:
            type: basic
            challenge: <em>false</em>
        authentication_backend:
            type: intern
    saml_auth_domain:
        http_enabled: true
        transport_enabled: false
        order: 1
        http_authenticator:
            type: saml
            challenge: true
            config:
                idp:

>                     metadata_file: "<em>/etc/wazuh-indexer/opensearch-security/wazuh_authentik_meta.xml</em>"

Reviewer 2: ```suggestion
                    metadata_file: "/etc/wazuh-indexer/opensearch-security/<em>wazuh_authentik_meta.xml</em>"
```

Since this dir is not to be changed

>                     entity_id: "<em>wazuh-saml</em>"

Reviewer 2: No need for `<em>`, right? On line something-somewhere-up-there, we tell the user to set a specific entity ID. as a result, wazuh-saml does not serve as a placeholder

                sp:
                    entity_id: "<em>wazuh-saml</em>"

>                 kibana_url: "<em>https://wazuh-dashboard.company/</em>"

Reviewer 2: ```suggestion
                kibana_url: "https://<em>wazuh-dashboard.company</em>/"
```

Sorry. Also, last I checked, doesn't the dashboard force SSL? so only wazuh-dashboard.company is a placeholder

                roles_key: Roles

>                 exchange_key: "<em><exchange key generated in step 1></em>"

Reviewer 2: hmm

        authentication_backend:
            type: noop
  `}
</IntegrationsMultilineCodeblock>
;

:::note

> Ensure that you set the following parameters in the `basic_internal_auth_domain` section: `order: 0` and `challenge: false`

Reviewer 2: so, if you say to set these to order:0 and challenge:false, why are they marked as placeholders with `<em>` above?

> And the `metadata_file`, `kibana_url`, and `exchange_key` parameters in the `saml_auth_domain` section.

Reviewer 2: Eh, they see the `<em>`. This isn't needed IMO

:::

> 5. Run the `securityadmin.sh` script to load the configuration changes made in the `/etc/wazuh-indexer/opensearch-security/config.yml` file:

Reviewer 2: ![image](https://github.com/user-attachments/assets/1c1fc329-ae32-487d-9015-564186b28b02)
![image](https://github.com/user-attachments/assets/dde45cc6-03ab-41f2-a365-e9b466cb8ddb)

Only 2 references of these files and they are right here. Did you mean to name that file differently?

Reviewer 2: Or am I missing something? It's getting late here

    ```bash
    export JAVA_HOME=/usr/share/wazuh-indexer/jdk/ && bash /usr/share/wazuh-indexer/plugins/opensearch-security/tools/securityadmin.sh -f /etc/wazuh-indexer/opensearch-security/config.yml -icl -key /etc/wazuh-indexer/certs/admin-key.pem -cert /etc/wazuh-indexer/certs/admin.pem -cacert /etc/wazuh-indexer/certs/root-ca.pem -h wazuh-indexer.company -nhnv
    ```

:::note

> The -h flag specifies the hostname or IP address of the Wazuh Indexer server.

Reviewer 2: TBH, we could `<kbd>` and `<em>` it and get rid of this note

:::

6. Edit the `/etc/wazuh-indexer/opensearch-security/roles_mapping.yml` file and include `wazuh_admin` as a `backend_role` in the appropriate section.

> for example in the `all_access` section:

Reviewer 2: hmmm

Author: I've updated this

    - "admin"
hosts: []
users: []
and_backend_roles: []
description: "Maps admin to all_access"
  `}
</IntegrationsMultilineCodeblock>
;

7. Run the `securityadmin.sh` script again but with the `-f` flag set to `/etc/wazuh-indexer/opensearch-security/roles_mapping.yml`:

    ```bash
    export JAVA_HOME=/usr/share/wazuh-indexer/jdk/ && bash /usr/share/wazuh-indexer/plugins/opensearch-security/tools/securityadmin.sh -f /etc/wazuh-indexer/opensearch-security/roles_mapping.yml -icl -key /etc/wazuh-indexer/certs/admin-key.pem -cert /etc/wazuh-indexer/certs/admin.pem -cacert /etc/wazuh-indexer/certs/root-ca.pem -h wazuh-indexer.company -nhnv
    ```

:::note
The -h flag specifies the hostname or IP address of the Wazuh Indexer server.
:::

### Wazuh dashboard configuration

1. On the Wazuh Dashboard server, check the value of `run_as` in the `/usr/share/wazuh-dashboard/data/wazuh/config/wazuh.yml` configuration file:

<IntegrationsMultilineCodeblock>
    {`
hosts:
    - default:
            url: https://127.0.0.1
            port: 55000
            username: wazuh-wui
            password: "<WAZUH_WUI_PASSWORD>"
            run_as: false
  `}
</IntegrationsMultilineCodeblock>
;

> If `run_as` is set to `false`, proceed to step 4.

Reviewer 2: hmmmm

Author: I've updated this

If `run_as` is set to `true`, you need to add a role mapping on the Wazuh dashboard:

2.  On the Wazuh dashboard click the upper-left menu icon ☰ to open the menu, go to **Server management** -> **Security** -> **Roles mapping**

3.  Click **Create Role mapping** and configure the following parameters:

    - **Role Name**: set a name for the role mapping (e.g. authentik_admins)
    - **Roles**: select a role (e.g.`administrator`)
    - **Custom rules**: Click **Add new rule** and set:
        - **User field**: `backend_roles`
        - **Search operation**: `FIND`
        - **Value**: `wazuh-admin`

Click **Save role mapping**

4. On the Wazuh Dashboard server, add these lines to the `/etc/wazuh-dashboard/opensearch_dashboards.yml` file:

<IntegrationsMultilineCodeblock>
    {`
opensearch_security.auth.type: "saml"
server.xsrf.allowlist:
    [
        "/_opendistro/_security/saml/acs",
        "/_opendistro/_security/saml/logout",
        "/_opendistro/_security/saml/acs/idpinitiated",
    ]
opensearch_security.session.keepalive: false
  `}

> </IntegrationsMultilineCodeblock>

Reviewer 2: ```suggestion
</IntegrationsMultilineCodeblock>;
```

> ;

Reviewer 2: ```suggestion
```

> 5. Restart the Wazuh dashboard service using this command:

Reviewer 2: IMO `the following command:` sounds better. What do you think @tanberry ?

    ```bash

>     systemctl restart wazuh-dashboard

Reviewer 2: sudo or nah
